### DISCUSSION TITLE
Time-Delayed Variables

### DISCUSSION THREAD

**Discussion by TadeuNP** [MilesCranmer/PySR#275] [Q&A]
Hi, 

I'm trying to use PySR to identify some equations with a certain $\tau$ dependence on the past, similar to the following delayed differential equation,

$$\begin{align}
    \frac{\mathrm{d}x}{\mathrm{d}t} = \frac{\beta}{1+(x(t-\tau))^n} - x.
\end{align}$$

One way of doing it is by introducing a new column with shifted data, but I was hoping for a method that allowed PySR itself to look for $\tau$. I thought about defining a binary operator B(x, tau) that returns the value of x at tau time steps ago, but looking at the file that defines the operators, it seems they only work with a single scalar value as input, so it looks incompatible.

Is there a way? Any suggestion? 

Thanks.

**Comment by MilesCranmer** [comment]
Hi @TadeuNP,

Very interesting idea. I think implementing this really depends on specific requirements. If you just want to query x in the past, you could certainly create a unary operator that takes a scalar argument, converts it to integer inside the function (type conversion is allowed - see the prime number example in the docs page on code examples), and indexes the correct row. However, all operators assume they are applied element-wise without changing functional form, so if you need the row index of the x, you will need to do some hacking.

~~If what you mean is a *functional*, rather than a numerical operator (ie the operator should change all of the variables in the subtree), that is not supported and would take a large amount of work.~~ (edit - see comment)

Cheers,
Miles

**Reply by MilesCranmer** [reply]
Actually I spoke too quickly. Implementing a functional isn’t so bad and I think you could actually do this. The latest backend version allows you to specify an arbitrary loss function — inside that loss function, you could apply your time-delay functional to the tree, converting features accordingly.

**Reply by MilesCranmer** [reply]
Added to PySR too: https://github.com/MilesCranmer/PySR/pull/276. Not sure how you could implement this specifically but it's probably possible.

**Reply by TadeuNP** [reply]
Hi @MilesCranmer, 

I saw your new addition to PySR and your comments about using an arbitrary loss function, but I'm not sure if I completely understand your suggestions.

I will briefly restate my situation to make sure we are talking about the same thing. Let's say we only have two features, $x_1$ and $x_2$, and they somehow combine to generate $y$. In the usual case, the nth value of $y$ should depend on some combination of $x_1^n$ and $x_2^n$ (I'm using an upper index to denote the row). In my case, $y^n$ actually depends on some past values of the features (e.g., $y^n = f(x_1^n, x_2^{n-3}$), but the exact choice of variable and how far in the past is something that needs to be discovered. 

Is it correct that your loss function suggestion is one way to have access to the whole dataset, not only to the nth value of a certain feature $x_k$ when evaluating the formula? So I would still need to introduce some kind of operator that selects the pair variable/tau, and when included in a tree, the custom loss function would somehow handle it (since it has access to all the dataset) and only then generate a prediction. 

Thanks.

**Reply by MilesCranmer** [reply]
Thanks. This is a bit difficult because operators don’t have access to that information. You can see in the call here: https://github.com/SymbolicML/DynamicExpressions.jl/blob/34271cb68c7286d90fe5600d6df173e5c05381bc/src/EvaluateEquation.jl#L153 - since the index “j” is not being passed to the operator, I’m not sure you could set this up easily.

Probably the easiest solution is to simply pass shifted variables as new columns. It might just work!

An alternative is to do some kind of symbolic distillation: https://speakerdeck.com/milescranmer/symbolic-distillation-of-neural-networks, where the NN is able to select the time delay. This requires more work on your end though.

**Reply by MilesCranmer** [reply]
Actually one way to get it to work is set up GenericOperatorEnum (https://github.com/SymbolicML/DynamicExpressions.jl/tree/34271cb68c7286d90fe5600d6df173e5c05381bc#generic-types) to work with SymbolicRegression.jl. Then the operator would actually have access to the entire array at once.

**Reply by TadeuNP** [reply]
I will take a look at all your new suggestions, but let me remark that it indeed works if I simply introduce shifted variables, and the most important thing is that PySR feature selection seems to be quite good and fast (I had not tried it before), so the problem of having too many shifted variables may not be that problematic after all!

edit: okay, the feature selection part is not working as well as it seemed to, but even with ~20 features it can still recover the expression if simple enough :)

**Reply by MilesCranmer** [reply]
Awesome!! Do you mean you are using `select_k_features`? Interestingly I actually don't usually use this feature - I find that just letting the GA run for a bit is enough to select the important features. I think the crossovers are good at doing the feature selection.

(I should really turn off the warning for >10 features)

**Reply by TadeuNP** [reply]
Yes, I was just testing it for the first time, but it seems to struggle with discriminating between very similar features , as is the case when $x_2$ is just a small shift of $x_1$. 

As you mentioned, it turns out that just running for a bit longer is enough to find the correct variables, even when I had 15+ delayed features introduced.
